libraries/chain/abi_serializer.cpp

```diff
@@ -160,18 +164,18 @@ namespace eosio { namespace chain {
    }
 
    bool abi_serializer::is_array(const type_name& type)const {
-      return ends_with(string(type), "[]");
+      return ends_with(string(type), array_mark);
    }
 
    bool abi_serializer::is_optional(const type_name& type)const {
-      return ends_with(string(type), "?");
+      return ends_with(string(type), optional_mark);
    }
 
    type_name abi_serializer::fundamental_type(const type_name& type)const {
       if( is_array(type) ) {
-         return type_name(string(type).substr(0, type.size()-2));
+         return type_name(string(type).substr(type.size()-array_mark.size()));
       } else if ( is_optional(type) ) {
-         return type_name(string(type).substr(0, type.size()-1));
+         return type_name(string(type).substr(type.size()-optional_mark.size()));
```

Comment:
This is not equivalent to the old code and is causing tests to fail. The first arg to `substr` is start position, second arg is length.

Reply:
yes, it is the wrong way.